# AI & ML Course with BinX — Week 1 — Day 4 — Pandas: Tabular Data 

## 4.1 Series and DataFrames


Quick example showing the difference between a Series and a DataFrame

In [8]:
import pandas as pd 
# Series : Single labeled column 
series = pd.Series([1, 2, 3], name="Example")
print(series) 
print(type(series))

# DataFrame : full table (columns and rows)
dataframe = pd.DataFrame({"even":[2, 4, 6], "odd":[1, 3, 5]})
print(dataframe)
print(type(dataframe))



0    1
1    2
2    3
Name: Example, dtype: int64
<class 'pandas.Series'>
   even  odd
0     2    1
1     4    3
2     6    5
<class 'pandas.DataFrame'>


## 4.2 Loading and Inspecting Data

In [13]:
df = pd.read_csv("data.csv")
df.head()

,country_code,country_name,year,value
0,ABW,Aruba,1960,11092.675
1,ABW,Aruba,1961,11576.719
2,ABW,Aruba,1962,12713.489
3,ABW,Aruba,1963,12178.107
4,ABW,Aruba,1964,11840.743


In [ ]:
df.shape        

(13953, 4)

In [ ]:
df.info()        

<class 'pandas.DataFrame'>
RangeIndex: 13953 entries, 0 to 13952
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   country_code  13953 non-null  str    
 1   country_name  13953 non-null  str    
 2   year          13953 non-null  int64  
 3   value         13953 non-null  float64
dtypes: float64(1), int64(1), str(2)
memory usage: 436.2 KB


In [16]:
df.describe()

,year,value
count,13953.000000,1.395300e+04
mean,1990.732316,8.254983e+05
std,17.187585,2.788923e+06
min,1960.000000,-8.067400e+01
25%,1976.000000,1.100000e+03
50%,1992.000000,1.390000e+04
75%,2006.000000,1.642779e+05
max,2019.000000,3.434401e+07


## 4.3 Selecting and Filtering

select single column

In [ ]:
df.columns

Index(['country_code', 'country_name', 'year', 'value'], dtype='str')

select multiple columns

In [22]:
df[["country_code", "country_name", "value"]]

,country_code,country_name,value
0,ABW,Aruba,11092.675000
1,ABW,Aruba,11576.719000
2,ABW,Aruba,12713.489000
3,ABW,Aruba,12178.107000
4,ABW,Aruba,11840.743000
...,...,...,...
13948,ZWE,Zimbabwe,12430.000305
13949,ZWE,Zimbabwe,11020.000458
13950,ZWE,Zimbabwe,10340.000153
13951,ZWE,Zimbabwe,12380.000114


Check The range of Value first to choose logical threshold


In [25]:
df["value"].describe

<bound method NDFrame.describe of 0        11092.675000
1        11576.719000
2        12713.489000
3        12178.107000
4        11840.743000
             ...     
13948    12430.000305
13949    11020.000458
13950    10340.000153
13951    12380.000114
13952    11760.000229
Name: value, Length: 13953, dtype: float64>

Boolean Filtering (rows where value is above thw maen)

In [26]:
threshold = df["value"].mean()
df[df["value"]>threshold]

,country_code,country_name,year,value
423,ARB,Arab World,1996,8.436291e+05
424,ARB,Arab World,1997,8.847770e+05
425,ARB,Arab World,1998,8.897156e+05
426,ARB,Arab World,1999,9.016245e+05
427,ARB,Arab World,2000,9.391917e+05
...,...,...,...,...
13647,WLD,World,2015,3.299554e+07
13648,WLD,World,2016,3.301856e+07
13649,WLD,World,2017,3.351454e+07
13650,WLD,World,2018,3.428935e+07


Using .loc  (country_name and year for rows above threshold)

In [27]:
df.loc[df["value"]>threshold, ["country_name","year"]]

,country_name,year
423,Arab World,1996
424,Arab World,1997
425,Arab World,1998
426,Arab World,1999
427,Arab World,2000
...,...,...
13647,World,2015
13648,World,2016
13649,World,2017
13650,World,2018


## 4.4 Cleaning Data

count missing values per column

In [28]:
df.isnull().sum()   

country_code    0
country_name    0
year            0
value           0
dtype: int64

There is no missing values founded but this is how you would handle them if there were _

**fill with mean**

df["value"].fillna(df["value"].mean(),inplace=True)  

**drop rows with any missing value**

df.dropna(inplace=True)        

print("No missing value found in this dataset")

In [30]:
print("Number of duplicate rows :", df.duplicated().sum())

Number of duplicate rows : 0


Remove Duplicates if any exist 

In [31]:
df.drop_duplicates(inplace=True)
print("Dataset after remove duplicates - Shape-",df.shape)

Dataset after remove duplicates - Shape- (13953, 4)


## 4.5 Grouping and Aggregation

Avarage CO2 Value per Cuntry

In [33]:
df.groupby("country_name")["value"].mean()


country_name
Afghanistan                    2.505611e+03
Africa Eastern and Southern    3.463568e+05
Africa Western and Central     1.075567e+05
Albania                        4.413947e+03
Algeria                        7.031805e+04
                                   ...     
West Bank and Gaza             1.754320e+03
World                          2.195928e+07
Yemen, Rep.                    8.619050e+03
Zambia                         3.369701e+03
Zimbabwe                       1.116555e+04
Name: value, Length: 256, dtype: float64

Multiple aggregation per countery  (mean and MAX value)

In [36]:
df.groupby("country_name").agg({"value": ["mean", "max"]})

value              
                                     mean           max
country_name                                           
Afghanistan                  2.505611e+03  8.930000e+03
Africa Eastern and Southern  3.463568e+05  6.030300e+05
Africa Western and Central   1.075567e+05  2.207400e+05
Albania                      4.413947e+03  8.984150e+03
Algeria                      7.031805e+04  1.712500e+05
...                                   ...           ...
West Bank and Gaza           1.754320e+03  3.234294e+03
World                        2.195928e+07  3.434401e+07
Yemen, Rep.                  8.619050e+03  2.783000e+04
Zambia                       3.369701e+03  7.730000e+03
Zimbabwe                     1.116555e+04  1.847000e+04

[256 rows x 2 columns]

# Hands-On Lab: Exploring a Real Dataset

Step 1: Load a CSV dataset and print shape, columns, dtypes

In [37]:
lab_df = pd.read_csv("data.csv")
print("Shape : ", lab_df.shape)
print("Columns : ", lab_df.columns.tolist())
print("Data Types :",lab_df.dtypes )

Shape :  (13953, 4)
Columns :  ['country_code', 'country_name', 'year', 'value']
Data Types : country_code        str
country_name        str
year              int64
value           float64
dtype: object


Step 2: Count missing values per column and handle them (with a documented justification)


In [38]:
missing_counts = lab_df.isnull().sum()
print("Missing values per column : ",missing_counts)

Missing values per column :  country_code    0
country_name    0
year            0
value           0
dtype: int64


**Justification:** The dataset has zero missing values in all columns
(country_code, country_name, year, value). No filling or dropping is needed —
the data is already complete and ready for analysis.

Step 3: Filter the data to a meaningful subset and describe what you find


In [40]:
lab_threshold = lab_df["value"].mean()
high_emission = lab_df[lab_df["value"]>lab_threshold]
print("Threshold(mean value) : ", lab_threshold)
print("Number of rows above threshold :",lab_df.shape[0])
high_emission.head()

Threshold(mean value) :  825498.3433543197
Number of rows above threshold : 13953


,country_code,country_name,year,value
423,ARB,Arab World,1996,843629.050531
424,ARB,Arab World,1997,884777.026039
425,ARB,Arab World,1998,889715.587754
426,ARB,Arab World,1999,901624.461624
427,ARB,Arab World,2000,939191.728434


In [44]:
print("Total rows in lab_df:", lab_df.shape[0])
print("Rows above threshold:", high_emission.shape[0])

Total rows in lab_df: 13953
Rows above threshold: 1727


**Findings:** Filtering the dataset to records above the mean value shows
1727 out of 13953 total rows represent higher-than-average CO2 emissions.
This subset highlights which countries/years contribute most to global emissions,
with entities like "Arab World" appearing among the higher-emission records.

Step 4: Use groupby to compute an aggregate statistic per category and interpret it in a Markdown cell


In [42]:
country_avg = lab_df.groupby("country_name")["value"].mean().sort_values(ascending=False)
country_avg.head(10)

country_name
World                        2.195928e+07
High income                  1.112143e+07
IDA & IBRD total             1.111222e+07
OECD members                 1.089645e+07
IBRD only                    1.066460e+07
Low & middle income          1.065039e+07
Post-demographic dividend    1.061569e+07
Middle income                1.049463e+07
Upper middle income          8.038619e+06
Late-demographic dividend    7.788907e+06
Name: value, dtype: float64